protocol/superchain-erc20/storage-upgrade.md

```diff
@@ -0,0 +1,133 @@
+
+To execute this path, it is necessary to modify the `OptimismMintableERC20Factory` contract to handle the `hashOnion` logic.
+
+Note: a similar implementation could be achieved with a Merkle root instead of a hash onion. Verifying and storing is easier in a hash onion, as it provides a clear ordering (automatically erases duplicates). On the other hand, a user who wants to convert with a legacy token will need to first unlock all previous layers in a hash onion, whereas in a Merkle tree, proofs are independent.
+
+## Tradeoffs and conclusion
+Pros of hash onion compared to full storage update:
+- requires a single implementation change in the proxy contract instead of the two necessary for the full storage upgrade.
+- doesn't require to precompute the storage slot for each element of the `deployments` mapping. This step should be straightforward, but any mistake requires a new upgrade.
+- easier to check a single value per chain compared to a token list per chain. 
+- fewer variables and, therefore, risk in the upgrade process.
+
+Cons of hash onion compared to full storage update:
+- after the `deployments` mapping is fully populated, the factory will have useless code. This can later be removed in a subsequent upgrade.
+- it requires user interaction. In particular, users might be unable to convert a given token unless its address has already been proven in the factory. This action might be reserved for power users.
+- less contract code, reducing on-chain risk.
+
+All things considered, the hash onion approach optimizes the upgrade process's security and scalability. Unused code should be a minor downside, which can later be addressed in future upgrades. Solid documentation and UI tools can minimize the risk of users getting stuck. 
+
+## Hash onion implementation details
+The idea of the hash onion is to commit to a data set with a single `bytes32` value. There are multiple ways to do so. In what follows, a simple reference implementation will be presented.
+
+Given a dataset `[d1, d2, ..., dn]`, the `hashOnion` will be built with the subsequent hashing of each data and the previous *layer* as follows:
+```
+layer0 = keccak256(d1)
+layer1 = keccak256(abi.encodePacked(layer0, d2))
+...
+hashOnion = layerN = keccak256(abi.encodePacked(layer(N-1), dN))
+```
+
+This operation compromises $N$ hashes. 
+The contract update will store the final `hashOnion`, and users will call a public function to reconstruct the whole dataset.
+
+Example implementation:
+```solidity
+/**
+ * @dev Verifies the previous layer and data, stores addresses in the mapping,
+ * and updates the last layer.
+ * @param localToken The local token address to be stored
+ * @param remoteToken The remote token address to be stored
+ * @param previousLayer The previous layer of the hash onion
+ */
+function verifyAndStore(
+	address localToken,
+	address remoteToken,
+	bytes32 previousLayer
```

Comment:
Could we rename `previousLayer` to `innerLayer`? "Previous" could be a bit confusing as it might also make reference to the layer already peeled before the current one.